java-squid/src/main/java/org/sonar/java/se/SymbolicValueRelation.java

```diff
@@ -34,73 +38,130 @@ protected SymbolicValueRelation(SymbolicValue v1, SymbolicValue v2) {
     this.v2 = v2;
   }
 
-  public Boolean impliedBy(List<SymbolicValueRelation> knownRelations) {
+  @Override
+  public int hashCode() {
+    long bits = v1.id();
+    bits ^= v2.id() * 31;
+    bits ^= getClass().hashCode() * 31;
+    return ((int) bits) ^ ((int) (bits >> 32));
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (obj instanceof SymbolicValueRelation) {
+      return equalsRelation((SymbolicValueRelation) obj);
+    }
+    return super.equals(obj);
+  }
+
+  private boolean equalsRelation(SymbolicValueRelation relation) {
+    final int[] key1 = key();
+    final int[] key2 = relation.key();
+    if (key1[0] == key2[0] && key1[1] == key2[1]) {
+      return getClass().equals(relation.getClass());
+    }
+    return false;
+  }
+
+  private int[] key() {
+    if (v1.id() < v2.id()) {
+      return new int[] {v1.id(), v2.id()};
+    } else {
+      return new int[] {v2.id(), v1.id()};
+    }
+  }
+
+  @CheckForNull
+  public Boolean impliedBy(Collection<SymbolicValueRelation> knownRelations) {
+    return impliedBy(knownRelations, new HashSet<SymbolicValueRelation>());
+  }
+
+  @CheckForNull
+  public Boolean impliedBy(Collection<SymbolicValueRelation> knownRelations, Set<SymbolicValueRelation> usedRelations) {
+    if (knownRelations.isEmpty()) {
+      return null;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Null is returned but a "Boolean" is expected. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2447)